sql/core/src/main/scala/org/apache/spark/sql/execution/python/TransformWithStateInPandasStateServer.scala

```diff
@@ -195,11 +201,76 @@ class TransformWithStateInPandasStateServer(
             None
         }
         initializeStateVariable(stateName, schema, StateVariableType.ListState, ttlDurationMs)
+      case StatefulProcessorCall.MethodCase.UTILSCALL =>
+        handleStatefulProcessorUtilRequest(message.getUtilsCall)
       case _ =>
         throw new IllegalArgumentException("Invalid method call")
     }
   }
 
+  private[sql] def handleStatefulProcessorUtilRequest(message: UtilsCallCommand): Unit = {
+    message.getMethodCase match {
+      case UtilsCallCommand.MethodCase.ISFIRSTBATCH =>
+        if (!hasInitialState) {
+          // In physical planning, hasInitialState will always be flipped
+          // if it is not first batch
+          sendResponse(1)
+        } else {
+          sendResponse(0)
+        }
+
+      case UtilsCallCommand.MethodCase.GETINITIALSTATE =>
+        if (!hasInitialState || initDataIter == null || !initDataIter.hasNext) {
+          sendResponse(1)
+        } else {
+          sendResponse(0)
+
+          outputStream.flush()
+          val outputSchema = new StructType()
+            .add("key", BinaryType)
+            .add("initialState", initialStateSchema)
+          val arrowSchema = ArrowUtils.toArrowSchema(outputSchema, timeZoneId,
+            errorOnDuplicatedFieldNames, largeVarTypes)
+          val allocator = ArrowUtils.rootAllocator.newChildAllocator(
+            s"stdout writer for transformWithStateInPandas state socket", 0, Long.MaxValue)
+          val root = VectorSchemaRoot.create(arrowSchema, allocator)
+          val writer = new ArrowStreamWriter(root, null, outputStream)
+          val arrowStreamWriter =
+            new BaseStreamingArrowWriter(root, writer,
+              arrowTransformWithStateInPandasMaxRecordsPerBatch)
+
+          val keyStatePair = initDataIter.next()
```

Comment:
IIUC, we only consume the first item/key in the initDataIter and send it to the python worker?